src/bluesearch/database/article.py

```diff
@@ -123,7 +173,7 @@ def uid(self) -> Optional[str]:
             If at least one identifier exists, unique id is created. Otherwise,
             the returned uid is None.
         """
-        return generate_uid((self.pubmed_id, self.pmc_id, self.doi))
+        return generate_uid((self.pubmed_id, self.pmc_id, self.arxiv_id, self.doi))
```

Comment:
> So this change will effectively "invalidate" all the existing UID's we have generated up until now 

Yes, and indeed I had to change the hashes in our unit tests, because the previous ones were now invalid.
As you say it's not a huge deal right now, but we should definitely fix the design of the `uid` computation before starting the actual population of the database, otherwise it will be a mess.

Is there any other `id` we could think of using? Maybe the `bioRxiv` and `medRxiv` ones?
This discussion is related to https://github.com/BlueBrain/Search/issues/507 (so maybe @Stannislav  any idea on other cases when the `uid` could be `None`?), we should try to make sure that the `uid` can always be computed.

Reply:
> Also, since the `arxiv_id` will hold the version it can lead to us having multiple versions of the same paper in our database. I guess that is what we want, right?

It's a design choice, I went for including the version to clarify any ambiguity. See also what they say in the `arXiv` documentation [here](https://arxiv.org/help/arxiv_identifier) (in particular, what happens if you don't specify the version):
> The identifier `arXiv:YYMM.numbervV` provides a complete and unique citation for an arXiv article. Without the version number (e.g. `arXiv:YYMM.number`), the identifier refers to the most recent version of the article.

Reply:
I agree that we should make sure that the UID computation includes all relevant identifiers before we actually start populating the database. Again, I guess this could be addressed in a separate issue/PR.

Reply:
See https://github.com/BlueBrain/Search/issues/507 for follow-up discussion on this.

Reply:
Regarding potential `None`s, I think according to #507 the sources we currently have are all covered.

As we discussed, we can think about a fall-back strategy for external/random texts that always works, like hashing the article title/abstract or something like that. I guess this is one of the potential places where this could be implemented. Something like
```python
@property
def uid(self) -> str:
    try:
        uid = generate_uid((self.pubmed_id, self.pmc_id, self.arxiv_id, self.doi))
    raise ValueError:
        uid = hash_article(self)
    return uid
```

Reply:
@Stannislav  To avoid scope creep, I copied your comment to #507 and I suggest we keep discussing it there.